allensdk/brain_observatory/behavior/data_objects/licks.py

```diff
@@ -86,15 +85,17 @@ def from_nwb(cls, nwbfile: NWBFile) -> Optional["Licks"]:
         if 'licking' in nwbfile.processing:
             lick_module = nwbfile.processing['licking']
             licks = lick_module.get_data_interface('licks')
-
-            df = pd.DataFrame({
-                'timestamps': licks.timestamps[:],
-                'frame': licks.data[:]
-            })
+            timestamps = licks.timestamps[:]
+            frame = licks.data[:]
         else:
-            warnings.warn("This session "
-                          f"'{int(nwbfile.identifier)}' has no rewards data.")
-            return None
+            timestamps = []
```

Comment:
I think this else clause will never get executed. The reason is that the old code [skips](https://github.com/AllenInstitute/AllenSDK/blob/v2.12.4/allensdk/brain_observatory/behavior/session_apis/data_io/behavior_nwb_api.py#L103) writing licks and rewards to nwb if the dataframe is empty. But we forgot to add that check in `BehaviorSession` so it will always get written and the if clause will always return True.

Reply:
But Marina and Alex both experienced `from_nwb` giving them `BehaviorSessions` with `self._licks = None`. That is the but the provoked this ticket (#2239).

I wonder if this is a backwards compatibility issue. NWB files that were written before your refactor skipped writing licks to the NWB file, so the clause executed. If we were to regenerate the dataset today, licks would always get written to the NWB files, even if they were just written as empty dataframes.

Reply:
Exactly, these nwb files were generated using the old code, and so they don't contain `Licks` or `Rewards`. But if we were to regenerate then `Licks` and `Rewards` would get written to NWB as empty dataframes. If we need to maintain backwards compatibility, then they should not get written to nwb.

Reply:
Is that important?

Whether or not licks and rewards get written to the NWB file, the user experience (mediated through `BehaviorSession`) is that they receive an empty dataframe for a passive session.

I expanded my validation script to include a round-tripping of the NWB file

```
import tempfile
from allensdk.brain_observatory.behavior.behavior_project_cache.\
    behavior_project_cache import \
        VisualBehaviorOphysProjectCache as cache_class

from allensdk.brain_observatory.behavior.behavior_ophys_experiment import (
    BehaviorOphysExperiment)

import allensdk

print(f'allensdk version {allensdk.__version__}')

dummy_dir = '/Users/scott.daniel/Pika/garage/'

exp_id=1010092804

cache_dir = tempfile.mkdtemp(prefix=dummy_dir)
cache = cache_class.from_s3_cache(cache_dir=cache_dir)
dataset = cache.get_behavior_ophys_experiment(exp_id)
licks = dataset.licks
print(licks)
rewards = dataset.rewards
print(rewards)
```

results in
```
allensdk version 2.13.1
ophys_session_table.csv: 100%|██████████████| 227k/227k [00:00<00:00, 3.04MMB/s]
behavior_session_table.csv: 100%|█████████| 1.21M/1.21M [00:00<00:00, 7.28MMB/s]
ophys_experiment_table.csv: 100%|███████████| 610k/610k [00:00<00:00, 4.87MMB/s]
ophys_cells_table.csv: 100%|██████████████| 4.29M/4.29M [00:00<00:00, 8.52MMB/s]
behavior_ophys_experiment_1010092804.nwb: 100%|█| 239M/239M [00:21<00:00, 11.0MM
Empty DataFrame
Columns: [timestamps, frame]
Index: []
Empty DataFrame
Columns: [volume, timestamps, autorewarded]
Index: []

round tripping

Empty DataFrame
Columns: [timestamps, frame]
Index: []
Empty DataFrame
Columns: [volume, timestamps, autorewarded]
Index: []
```
So, from the user's perspective, there's no difference between the old-style NWB file and the new-style NWB file.

Reply:
So that we are on the same page -- the else block will only get executed if loading an nwb file generated with old code. But if we were to generate a new nwb file with the current code, it will never get executed. I can't remember -- is that a use case we support? Using the cache with a version of AllenSDK that is different from the one used to generate the NWB files?

Reply:
I've added the check we talked about to `to_nwb`